make/config.mk

```diff
@@ -132,10 +132,19 @@ endif
 ARCH := $(shell uname -a)
 ifneq (,$(filter $(ARCH), armv6l armv7l powerpc64le ppc64le aarch64))
 	USE_SSE=0
+	USE_F16C=0
 else
 	USE_SSE=1
 endif
 
+#----------------------------
+# F16C instruction support for faster arithmetic of fp16 on CPU
+#----------------------------
+# For distributed training with fp16, this helps even if training on GPUs
+# If left empty, checks CPU support and turns it on.
+# For cross compilation, please check support for F16C on target device and turn off if necessary.
+USE_F16C =
```

Comment:
Why aren't we adding similar logic like CMake for F16C support ?

Reply:
That is in mshadow's makefile. I didn't add it here because it is always included in all code paths.

Reply:
IMO, we should try to do most work (including whether hardware has FP16 support, and compiler has Fp16 support) in MXNet itself and pass the flag to Mshadow, so that we reduce the dependency on mshadow. This way MXNet users won't be impacted if something in MShadow breaks. WDYT ?

Reply:
I think this is hard to pull off, considering it's a compile flag, and overcomplicates our build-process even further. As @hen pointed out on dev@, we should rather find a more permanent solution rather than trying to add even more hacks

Reply:
@rahul003 @marcoabreu I am okay with leaving this for now if its too much effort. Having said that, I think this stop-gap solution is better than what we have today if it means less pain for the user until we move to the more permanent solution. Considering that permanent solution (or atleast one of them) would be to move all the dmlc projects under apache (so that MXNet has better control on the repos), I don't see it happening anytime soon.

Reply:
I've added the logic of identifying the flag here in MXNet. MShadow's role now is to use that flag if defined and set the compiler flag. This means we have control over when to turn on or off the flag in MXNet